modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Apply.kt

```diff
@@ -38,6 +38,12 @@ interface Apply<F> : Functor<F> {
    */
   fun <A, B> Kind<F, A>.ap(ff: Kind<F, (A) -> B>): Kind<F, B>
 
+  /**
+   * Lazy version of ap, useful for datatypes which can short circuit.
+   * This will at some point be removed for a compiler plugin
+   */
+  fun <A, B> Kind<F, A>.lazyAp(ff: () -> Kind<F, (A) -> B>): Kind<F, B> = ap(ff())
```

Comment:
The default implementation is just `ap` because most applicatives don't benefit from `lazyAp` only those with a short-circuit pattern do and forcing those to implement a pointless method is more work than having the others explicitly implement it